{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################\n",
    "######################################################## cell type relationship inference using cell type fraction\n",
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "import matplotlib\n",
    "matplotlib.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "from matplotlib import rc\n",
    "from scipy import stats\n",
    "import itertools\n",
    "from scipy import linalg\n",
    "from sklearn import mixture\n",
    "######################################## read metadata for IPF single cell RNA-Seq data\n",
    "path = 'E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\'\n",
    "x = pd.read_csv(path+'GSE136831_AllCells.Samples.CellType.MetadataTable.txt',sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(312928, 9)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CellBarcode_Identity</th>\n",
       "      <th>nUMI</th>\n",
       "      <th>nGene</th>\n",
       "      <th>CellType_Category</th>\n",
       "      <th>Manuscript_Identity</th>\n",
       "      <th>Subclass_Cell_Identity</th>\n",
       "      <th>Disease_Identity</th>\n",
       "      <th>Subject_Identity</th>\n",
       "      <th>Library_Identity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>001C_AAACCTGCATCGGGTC</td>\n",
       "      <td>5477</td>\n",
       "      <td>2150</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>ncMonocyte</td>\n",
       "      <td>Monocyte_Non-Classical</td>\n",
       "      <td>Control</td>\n",
       "      <td>001C</td>\n",
       "      <td>001C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>001C_AAACCTGTCAACACCA</td>\n",
       "      <td>20311</td>\n",
       "      <td>4726</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Macrophage_Alveolar</td>\n",
       "      <td>Macrophage_Alveolar</td>\n",
       "      <td>Control</td>\n",
       "      <td>001C</td>\n",
       "      <td>001C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>001C_AAACCTGTCACAGTAC</td>\n",
       "      <td>1390</td>\n",
       "      <td>881</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK</td>\n",
       "      <td>Control</td>\n",
       "      <td>001C</td>\n",
       "      <td>001C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>001C_AAACCTGTCTGTCTAT</td>\n",
       "      <td>3968</td>\n",
       "      <td>1943</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>cMonocyte</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Control</td>\n",
       "      <td>001C</td>\n",
       "      <td>001C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>001C_AAACGGGAGACTAAGT</td>\n",
       "      <td>3036</td>\n",
       "      <td>1716</td>\n",
       "      <td>Endothelial</td>\n",
       "      <td>Lymphatic</td>\n",
       "      <td>Lymphatic-Endothelial</td>\n",
       "      <td>Control</td>\n",
       "      <td>001C</td>\n",
       "      <td>001C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>001C_AAACGGGAGGCTCATT</td>\n",
       "      <td>4514</td>\n",
       "      <td>1936</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Macrophage</td>\n",
       "      <td>Macrophage</td>\n",
       "      <td>Control</td>\n",
       "      <td>001C</td>\n",
       "      <td>001C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>001C_AAACGGGAGGGAACGG</td>\n",
       "      <td>3773</td>\n",
       "      <td>1355</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>cMonocyte</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Control</td>\n",
       "      <td>001C</td>\n",
       "      <td>001C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>001C_AAACGGGCAGTAGAGC</td>\n",
       "      <td>2135</td>\n",
       "      <td>991</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>cMonocyte</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Control</td>\n",
       "      <td>001C</td>\n",
       "      <td>001C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>001C_AAACGGGGTATAATGG</td>\n",
       "      <td>7569</td>\n",
       "      <td>2890</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Macrophage_Alveolar</td>\n",
       "      <td>Macrophage_Alveolar</td>\n",
       "      <td>Control</td>\n",
       "      <td>001C</td>\n",
       "      <td>001C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>001C_AAACGGGGTTTAGGAA</td>\n",
       "      <td>5755</td>\n",
       "      <td>2760</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Macrophage</td>\n",
       "      <td>Macrophage_CellCycle_B</td>\n",
       "      <td>Control</td>\n",
       "      <td>001C</td>\n",
       "      <td>001C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    CellBarcode_Identity   nUMI  nGene CellType_Category  Manuscript_Identity  \\\n",
       "0  001C_AAACCTGCATCGGGTC   5477   2150           Myeloid           ncMonocyte   \n",
       "1  001C_AAACCTGTCAACACCA  20311   4726           Myeloid  Macrophage_Alveolar   \n",
       "2  001C_AAACCTGTCACAGTAC   1390    881          Lymphoid                   NK   \n",
       "3  001C_AAACCTGTCTGTCTAT   3968   1943           Myeloid            cMonocyte   \n",
       "4  001C_AAACGGGAGACTAAGT   3036   1716       Endothelial            Lymphatic   \n",
       "5  001C_AAACGGGAGGCTCATT   4514   1936           Myeloid           Macrophage   \n",
       "6  001C_AAACGGGAGGGAACGG   3773   1355           Myeloid            cMonocyte   \n",
       "7  001C_AAACGGGCAGTAGAGC   2135    991           Myeloid            cMonocyte   \n",
       "8  001C_AAACGGGGTATAATGG   7569   2890           Myeloid  Macrophage_Alveolar   \n",
       "9  001C_AAACGGGGTTTAGGAA   5755   2760           Myeloid           Macrophage   \n",
       "\n",
       "   Subclass_Cell_Identity Disease_Identity Subject_Identity Library_Identity  \n",
       "0  Monocyte_Non-Classical          Control             001C             001C  \n",
       "1     Macrophage_Alveolar          Control             001C             001C  \n",
       "2                      NK          Control             001C             001C  \n",
       "3                Monocyte          Control             001C             001C  \n",
       "4   Lymphatic-Endothelial          Control             001C             001C  \n",
       "5              Macrophage          Control             001C             001C  \n",
       "6                Monocyte          Control             001C             001C  \n",
       "7                Monocyte          Control             001C             001C  \n",
       "8     Macrophage_Alveolar          Control             001C             001C  \n",
       "9  Macrophage_CellCycle_B          Control             001C             001C  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.iloc[0:10,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "################### rank cell types based on their counts\n",
    "x_cell_type_count = x['Manuscript_Identity'].value_counts()\n",
    "cell_types_whole = x_cell_type_count.index.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types_whole"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "################ for each individual, calculate its cell type fraction\n",
    "disease_state = ['Control','IPF']\n",
    "ctx_list = []\n",
    "for disease in disease_state:\n",
    "    x1 = x[x['Disease_Identity']==disease]\n",
    "    ctx = pd.crosstab(x1['Subject_Identity'], x1['Manuscript_Identity'], normalize='index')\n",
    "    ctx = ctx.fillna(0)\n",
    "    ctx_list.append(ctx)\n",
    "\n",
    "ctxx = pd.concat(ctx_list,axis = 0)\n",
    "# cty = pd.concat(ctx_list_list) ## 35 31 20\n",
    "cty = ctxx.fillna(0)\n",
    "cell_pro = pd.DataFrame(cty)\n",
    "cell_pro = cell_pro.fillna(0)\n",
    "disease_lis = pd.Series(np.array(['Control' for i in range(28)] + ['IPF' for i in range(32)]))\n",
    "cell_pro = cell_pro.assign(disease=disease_lis.values)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_pro.shape\n",
    "cell_pro.iloc[0:10,0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "############################################### for each cell type, using its fraction value to train one Guassian Mixture Model with two components for all individuals\n",
    "\n",
    "X = np.log10(cell_pro.iloc[:,0:-1]+10**-2)\n",
    "Y_comp = []\n",
    "n_components_range = range(2, 3)\n",
    "for cell_type in cell_types_whole:\n",
    "    print (cell_type)\n",
    "    lowest_bic = np.infty\n",
    "    bic = []\n",
    "    for n_components in n_components_range:\n",
    "        # Fit a Gaussian mixture with EM\n",
    "        gmm = mixture.GaussianMixture(n_components=n_components,max_iter=500,tol =1*10**-4)#,covariance_type='diag',tol =10**-5)#,means_init =np.array([[-1.5],[-0.5]]))\n",
    "        gmm.fit(np.array(X[cell_type])[:,np.newaxis])\n",
    "        bic.append(gmm.bic(np.array(X[cell_type])[:,np.newaxis]))\n",
    "        if bic[-1] < lowest_bic:\n",
    "            lowest_bic = bic[-1]\n",
    "            best_gmm = gmm\n",
    "    Y=best_gmm.predict(np.array(X[cell_type])[:,np.newaxis])\n",
    "    Y_comp.append(Y)\n",
    "\n",
    "f, axes = plt.subplots(6, 7,  sharex=True,figsize=(15, 15))\n",
    "for i in range(cell_pro.shape[1]-1):\n",
    "    sns.distplot( X[cell_types_whole[i]] , ax=axes[i//7, i%7])\n",
    "    for j in range(Y_comp[i].max()+1):\n",
    "        label_index = np.where(Y_comp[i]==j)[0]\n",
    "        ax= sns.distplot(X[cell_types_whole[i]].iloc[label_index], ax=axes[i // 7, i % 7],hist = False, rug=True, label =str(j))\n",
    "    # plt.setp(ax.get_legend().get_texts(), fontsize='2')  # for legend text\n",
    "    # plt.setp(ax.get_legend().get_title(), fontsize='2')  # for legend title\n",
    "\n",
    "plt.savefig(path+'\\cell_type_prop_distri_GMM_log_2components_jupyter.pdf',bbox_inches='tight')\n",
    "del f, axes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def get_GMM_cell_pro (cell_pro,cell_types,indexx,data_path):\n",
    "    X = np.log10(cell_pro.iloc[:,1:-1]+10**-2)\n",
    "    Y_comp = []\n",
    "    n_components_range = range(2, 3)\n",
    "    for cell_type in cell_types:\n",
    "        #print (cell_type)\n",
    "        lowest_bic = np.infty\n",
    "        bic = []\n",
    "        for n_components in n_components_range:\n",
    "            # Fit a Gaussian mixture with EM\n",
    "            gmm = mixture.GaussianMixture(n_components=n_components,max_iter=500,tol =1*10**-4)#,covariance_type='diag',tol =10**-5)#,means_init =np.array([[-1.5],[-0.5]]))\n",
    "            gmm.fit(np.array(X[cell_type])[:,np.newaxis])\n",
    "            bic.append(gmm.bic(np.array(X[cell_type])[:,np.newaxis]))\n",
    "            if bic[-1] < lowest_bic:\n",
    "                lowest_bic = bic[-1]\n",
    "                best_gmm = gmm\n",
    "        Y=best_gmm.predict(np.array(X[cell_type])[:,np.newaxis])\n",
    "        Y_comp.append(Y)\n",
    "    np.save(data_path+'\\GMM_comp_'+str(indexx)+'.npy',np.array(Y_comp))\n",
    "    h_index = {}\n",
    "    h_index[0]='\"0\"'\n",
    "    h_index[1]='\"1\"'\n",
    "    h_index[2]='\"2\"'\n",
    "    h_index[3]='\"3\"'\n",
    "    h_index[4]='\"4\"'\n",
    "    Z_label= np.zeros((X.shape[0],cell_types.shape[0]))\n",
    "    Z_label_df = pd.DataFrame(Z_label,columns  = cell_types )\n",
    "    for i in range(cell_types.shape[0]):\n",
    "        ratios =  X[cell_types[i]]\n",
    "        labels = Y_comp[i]\n",
    "        mean_ratios = []\n",
    "        for j in range (labels.max()+1):\n",
    "            mean_ratios.append(ratios.iloc[np.where(labels==j)[0]].mean())\n",
    "        sort_mean_ratios = np.argsort(mean_ratios)\n",
    "        for j in range(len(sort_mean_ratios)):\n",
    "            Z_label_df.iloc[np.where(labels==sort_mean_ratios[j])[0],i] = h_index[j]\n",
    "    Z_label_df = Z_label_df.assign(disease=cell_pro.iloc[:,-1].values)\n",
    "    Z_label_df.to_csv(data_path+'\\cell_pro_sort_discrete_'+str(indexx)+'.csv')\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################  bootsrap cells for 100 times\n",
    "import random\n",
    "import os\n",
    "folder = os.path.exists(path+'\\IPF_control_cell_sampling_no_COPD')\n",
    "if not folder:\n",
    "    os.makedirs(path+'\\IPF_control_cell_sampling_no_COPD')\n",
    "whole_cell_index = range(x.shape[0])\n",
    "num_selection = int(x.shape[0]*0.8)\n",
    "for i in range (100):\n",
    "    print (i)\n",
    "    random.seed(i)\n",
    "    ctx_list = []\n",
    "    random_selection= random.sample(whole_cell_index, num_selection)\n",
    "    y = x.iloc[random_selection,]\n",
    "    for disease in disease_state:\n",
    "        x1 = x[x['Disease_Identity'] == disease]\n",
    "        ctx = pd.crosstab(x1['Subject_Identity'], x1['Manuscript_Identity'], normalize='index')\n",
    "        ctx = ctx.fillna(0)\n",
    "        ctx_list.append(ctx)\n",
    "    ctxx = pd.concat(ctx_list,axis = 0)\n",
    "    # cty = pd.concat(ctx_list_list) ## 39 31 20\n",
    "    cty = ctxx.fillna(0)\n",
    "    cell_pro = pd.DataFrame(cty)\n",
    "    cell_pro = cell_pro.fillna(0)\n",
    "    disease_lis = pd.Series(\n",
    "        np.array(['Control' for i in range(28)] + ['IPF' for i in range(32)] ))\n",
    "    cell_pro = cell_pro.assign(disease=disease_lis.values)\n",
    "    cell_pro.to_csv (path+'\\IPF_control_cell_sampling_no_COPD\\cell_pro_'+str(i)+'.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = path+'\\IPF_control_cell_sampling_no_COPD_top_cell_types\\\\' ########## get 100 bootsrap GMM models\n",
    "folder = os.path.exists(data_path)\n",
    "if not folder:\n",
    "    os.makedirs(data_path)\n",
    "cell_types = cell_types_whole\n",
    "for i in range (100):\n",
    "    print (i)\n",
    "    cell_pro_rand = pd.read_csv (path+'\\IPF_control_cell_sampling_no_COPD\\cell_pro_'+str(i)+'.csv')\n",
    "    get_GMM_cell_pro(cell_pro_rand, cell_types, i, data_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "############### Using R to do Bayesian Network inferenece please run it using R or R studio R.3.6.1\n",
    "######################### update Yale data//control / top cell types\n",
    "#path = 'E:/backup/update_data_20201210/update_data_jupyter/IPF_control_cell_sampling_no_COPD_top_cell_types/'\n",
    "#dir.create(paste(path,'IPF_control_cell_sampling_no_COPD_top_cell_types/Control_edges_37',sep=''))\n",
    "#dir.create(\"E:/backup/update_data_20201210/update_data_jupyter/IPF_control_cell_sampling_no_COPD_top_cell_types/IPF_control_cell_sampling_no_COPD_top_cell_types/Control_edges_37\")\n",
    "#dir.create(\"E:/backup/update_data_20201210/update_data_jupyter/IPF_control_cell_sampling_no_COPD_top_cell_types/IPF_control_cell_sampling_no_COPD_top_cell_types/IPF_edges_37\")\n",
    "\n",
    "## please creat the dir by manual\n",
    "library(bnlearn)\n",
    "library(Rgraphviz)\n",
    "for (j in c(0:99))\n",
    "{\n",
    "  print (j)\n",
    "  data_path <- paste('E:/backup/update_data_20201210/update_data_jupyter/IPF_control_cell_sampling_no_COPD_top_cell_types/','cell_pro_sort_discrete_',j,'.csv',sep='')\n",
    "  data <-read.csv(file = data_path,header = T)\n",
    "  datax = data[1:28,2:38]\n",
    "  #d = discretize(data, method = 'hartemink', breaks = 3, ibreaks = 20)\n",
    "  res = rsmax2(datax)\n",
    "  write.table(res$arcs, file = paste('E:/backup/update_data_20201210/update_data_jupyter/IPF_control_cell_sampling_no_COPD_top_cell_types/Control_edges_37/cell_pro_discrete_result_',j,'.txt',sep=''), sep = \" \")\n",
    "}\n",
    "\n",
    "######################### update Yale data//IPF\n",
    "for (j in c(0:99))\n",
    "{\n",
    "  print (j)\n",
    "\n",
    "  data_path <- paste('E:/backup/update_data_20201210/update_data_jupyter/IPF_control_cell_sampling_no_COPD_top_cell_types/cell_pro_sort_discrete_',j,'.csv',sep='')\n",
    "  data <-read.csv(file = data_path,header = T)\n",
    "  datax = data[29:60,2:38]\n",
    "  #d = discretize(data, method = 'hartemink', breaks = 3, ibreaks = 20)\n",
    "  res = rsmax2(datax)\n",
    "  write.table(res$arcs, file = paste('E:/backup/update_data_20201210/update_data_jupyter/IPF_control_cell_sampling_no_COPD_top_cell_types/IPF_edges_37/cell_pro_discrete_result_',j,'.txt',sep=''), sep = \" \")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###################################Back to Python\n",
    "##################################  BN results analysis\n",
    "\n",
    "import os\n",
    "import networkx as nx\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from networkx.drawing.nx_agraph import graphviz_layout\n",
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "path = \"E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\IPF_control_cell_sampling_no_COPD_top_cell_types\\\\IPF_edges_37\" #\n",
    "files= os.listdir(path)\n",
    "h_cell_type_pair = {}\n",
    "for file in files: #\n",
    "     if not os.path.isdir(file): #\n",
    "        print (file)\n",
    "        f = open(path+\"/\"+file) #\n",
    "        next(f)\n",
    "        for line in f: #\n",
    "            info = line.strip().split()\n",
    "            # print (info)\n",
    "            cell_type_pair = info[1][1:-1]+'\\t'+info[2][1:-1]\n",
    "            if cell_type_pair not in h_cell_type_pair:\n",
    "                h_cell_type_pair[cell_type_pair] = 1\n",
    "            else:\n",
    "                h_cell_type_pair[cell_type_pair] = h_cell_type_pair[cell_type_pair]+1\n",
    "        f.close()\n",
    "\n",
    "h_cell_type_pair_IPF = h_cell_type_pair\n",
    "\n",
    "path = \"E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\IPF_control_cell_sampling_no_COPD_top_cell_types\\\\Control_edges_37\" #\n",
    "files= os.listdir(path)\n",
    "h_cell_type_pair = {}\n",
    "for file in files: #\n",
    "     if not os.path.isdir(file): #\n",
    "        print (file)\n",
    "        f = open(path+\"/\"+file) #\n",
    "        next(f)\n",
    "        for line in f: #\n",
    "            info = line.strip().split()\n",
    "            # print (info)\n",
    "            cell_type_pair = info[1][1:-1]+'\\t'+info[2][1:-1]\n",
    "            if cell_type_pair not in h_cell_type_pair:\n",
    "                h_cell_type_pair[cell_type_pair] = 1\n",
    "            else:\n",
    "                h_cell_type_pair[cell_type_pair] = h_cell_type_pair[cell_type_pair]+1\n",
    "        f.close()\n",
    "        \n",
    "h_cell_type_pair_control = h_cell_type_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_pair_ovlp = set(h_cell_type_pair_control.keys())| set(h_cell_type_pair_IPF.keys())\n",
    "dfx = defaultdict(list)\n",
    "for cell_type_pair in cell_type_pair_ovlp:\n",
    "    dfx['cell_type'].append(cell_type_pair)\n",
    "    if cell_type_pair in h_cell_type_pair_control:\n",
    "        dfx['control'].append(h_cell_type_pair_control[cell_type_pair])\n",
    "    else:\n",
    "        dfx['control'].append(0)\n",
    "    if cell_type_pair in h_cell_type_pair_IPF:\n",
    "        dfx['IPF'].append(h_cell_type_pair_IPF[cell_type_pair])\n",
    "    else:\n",
    "        dfx['IPF'].append(0)\n",
    "dfxx = pd.DataFrame(dfx)\n",
    "\n",
    "dfxx.to_csv('E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\IPF_control_cell_sampling_no_COPD_top_cell_types\\\\top_cell_types_37_IPF_vs_Control_cell_sampling_BDN_cell_type_pair_num.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "################ we then convert the csv file to a standard file that has the following columns\n",
    "\"cell_type1\" \"cell_type2\" \"control\" \"IPF\" \"IPF-control\" \"abs(IPF-control)\"\n",
    "dff=pd.read_csv('E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\IPF_control_cell_sampling_no_COPD_top_cell_types\\\\top_cell_types_37_IPF_vs_Control_cell_sampling_BDN_cell_type_pair_num.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_type1</th>\n",
       "      <th>cell_type2</th>\n",
       "      <th>control</th>\n",
       "      <th>IPF</th>\n",
       "      <th>diff</th>\n",
       "      <th>abs_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Macrophage</td>\n",
       "      <td>cDC1</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VE_Capillary_B</td>\n",
       "      <td>SMC</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>-100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mesothelial</td>\n",
       "      <td>Aberrant_Basaloid</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>DC_Mature</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ncMonocyte</td>\n",
       "      <td>Multiplet</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>VE_Peribronchial</td>\n",
       "      <td>Pericyte</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>cDC1</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>-100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Ciliated</td>\n",
       "      <td>ncMonocyte</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>-100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>B_Plasma</td>\n",
       "      <td>Mesothelial</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>-100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>Lymphatic</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>-100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Macrophage</td>\n",
       "      <td>Ciliated</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>-100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Macrophage_Alveolar</td>\n",
       "      <td>pDC</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>-100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Multiplet</td>\n",
       "      <td>VE_Capillary_B</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ncMonocyte</td>\n",
       "      <td>DC_Mature</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>-100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>T_Regulatory</td>\n",
       "      <td>DC_Langerhans</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>-100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Myofibroblast</td>\n",
       "      <td>VE_Venous</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>-100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Pericyte</td>\n",
       "      <td>SMC</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>T</td>\n",
       "      <td>T_Regulatory</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>VE_Arterial</td>\n",
       "      <td>VE_Venous</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>T_Regulatory</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Myofibroblast</td>\n",
       "      <td>Lymphatic</td>\n",
       "      <td>0</td>\n",
       "      <td>76</td>\n",
       "      <td>76</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>T_Cytotoxic</td>\n",
       "      <td>NK</td>\n",
       "      <td>61</td>\n",
       "      <td>0</td>\n",
       "      <td>-61</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ILC_A</td>\n",
       "      <td>ILC_B</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SMC</td>\n",
       "      <td>VE_Capillary_A</td>\n",
       "      <td>55</td>\n",
       "      <td>0</td>\n",
       "      <td>-55</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Lymphatic</td>\n",
       "      <td>Macrophage</td>\n",
       "      <td>0</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Lymphatic</td>\n",
       "      <td>T</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>-50</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>VE_Arterial</td>\n",
       "      <td>VE_Capillary_B</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>49</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Macrophage</td>\n",
       "      <td>Lymphatic</td>\n",
       "      <td>0</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>Myofibroblast</td>\n",
       "      <td>0</td>\n",
       "      <td>45</td>\n",
       "      <td>45</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>T</td>\n",
       "      <td>ILC_B</td>\n",
       "      <td>44</td>\n",
       "      <td>0</td>\n",
       "      <td>-44</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             cell_type1         cell_type2  control  IPF  diff  abs_diff\n",
       "0            Macrophage               cDC1        0  100   100       100\n",
       "1        VE_Capillary_B                SMC      100    0  -100       100\n",
       "2           Mesothelial  Aberrant_Basaloid        0  100   100       100\n",
       "3                  cDC2          DC_Mature        0  100   100       100\n",
       "4            ncMonocyte          Multiplet        0  100   100       100\n",
       "5      VE_Peribronchial           Pericyte        0  100   100       100\n",
       "6                  cDC2               cDC1      100    0  -100       100\n",
       "7              Ciliated         ncMonocyte      100    0  -100       100\n",
       "8              B_Plasma        Mesothelial      100    0  -100       100\n",
       "9            Fibroblast          Lymphatic      100    0  -100       100\n",
       "10           Macrophage           Ciliated      100    0  -100       100\n",
       "11  Macrophage_Alveolar                pDC      100    0  -100       100\n",
       "12            Multiplet     VE_Capillary_B        0  100   100       100\n",
       "13           ncMonocyte          DC_Mature      100    0  -100       100\n",
       "14         T_Regulatory      DC_Langerhans      100    0  -100       100\n",
       "15        Myofibroblast          VE_Venous      100    0  -100       100\n",
       "16             Pericyte                SMC        0  100   100       100\n",
       "17                    T       T_Regulatory        0  100   100       100\n",
       "18          VE_Arterial          VE_Venous        0  100   100       100\n",
       "19         T_Regulatory         Fibroblast        0  100   100       100\n",
       "20        Myofibroblast          Lymphatic        0   76    76        76\n",
       "21          T_Cytotoxic                 NK       61    0   -61        61\n",
       "22                ILC_A              ILC_B        0   57    57        57\n",
       "23                  SMC     VE_Capillary_A       55    0   -55        55\n",
       "24            Lymphatic         Macrophage        0   54    54        54\n",
       "25            Lymphatic                  T       50    0   -50        50\n",
       "26          VE_Arterial     VE_Capillary_B        0   49    49        49\n",
       "27           Macrophage          Lymphatic        0   46    46        46\n",
       "28           Fibroblast      Myofibroblast        0   45    45        45\n",
       "29                    T              ILC_B       44    0   -44        44"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dff.iloc[0:30,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#########################################################################################\n",
    "#########################################################################################\n",
    "################ next we do ligand receptor regression analysis for the cell type pairs generated by the Bayesian Network\n",
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "import matplotlib\n",
    "matplotlib.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "import random\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.linear_model import LassoCV, LassoLarsCV, LassoLarsIC\n",
    "from sklearn import datasets\n",
    "from scipy import stats\n",
    "import pickle\n",
    "import time\n",
    "import pandas as pd\n",
    "#########################################\n",
    "##\n",
    "# we use the ligand-target interaction matrix from NicheNet, and then calculated each cell type's ligand and target average expression\n",
    "# across all individuals in control case and in IPF case respectively. And  we first normalize the raw data so that the sum of expression of all genes in one cell is 1.\n",
    "### the absoulte value does not matter because we are interested in expression fold change between case and control.\n",
    "\n",
    "\n",
    "#ligand_target_matrix = pd.read_csv('E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\ligand_target_matrix.txt',sep=' ')\n",
    "# df_expx = pd.read_csv('E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\average_gene_exp_cell_type_IPF_Control.csv')\n",
    "#\n",
    "# ############ gene name alignment\n",
    "# ligand_list = list(ligand_target_matrix)\n",
    "# target_list = list(ligand_target_matrix.index)\n",
    "# ligand_list_lower = [i.lower() for i in ligand_list]\n",
    "# target_list_lower = [i.lower() for i in target_list]\n",
    "#\n",
    "# expression_data_gene_list = df_expx['gene'][0:45947] ## target_list contains ligand list is ok, since target is in another cell\n",
    "# h_expression_gene_index = {}\n",
    "# for i in range (len(expression_data_gene_list)):\n",
    "#     h_expression_gene_index[expression_data_gene_list[i]] = i\n",
    "#\n",
    "# target_in_target_index = []\n",
    "# target_in_expression_index = []\n",
    "# ligand_in_target_index = []\n",
    "# ligand_in_expression_index = []\n",
    "#\n",
    "# for i in range(len(target_list_lower)):\n",
    "#     if target_list_lower[i] in h_expression_gene_index:\n",
    "#         target_in_target_index.append(i)\n",
    "#         target_in_expression_index.append(h_expression_gene_index[target_list_lower[i]]) ### get target gene index\n",
    "#\n",
    "# for i in range(len(ligand_list_lower)):\n",
    "#     if ligand_list_lower[i] in h_expression_gene_index:\n",
    "#         ligand_in_target_index.append(i)\n",
    "#         ligand_in_expression_index.append(h_expression_gene_index[ligand_list_lower[i]])  ## get ligand gene index\n",
    "#\n",
    "# aligned_ligand_target_matrix = ligand_target_matrix.iloc[target_in_target_index,ligand_in_target_index]\n",
    "# aligned_ligand_target_matrix.to_csv('E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\ligand_target_matrix_aligned_yale.csv')\n",
    "#\n",
    "# aligned_df_exp_target_index = []\n",
    "# for i in range (39*2): ### 39 cell types \n",
    "#     aligned_df_exp_target_index = aligned_df_exp_target_index + list(np.array(target_in_expression_index)+45947*i)\n",
    "#\n",
    "# aligned_df_exp_target = df_expx.iloc[aligned_df_exp_target_index,]\n",
    "# aligned_df_exp_targetx = aligned_df_exp_target.fillna(0)\n",
    "# aligned_df_exp_targetx.to_csv('E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\target_expression_aligned_yale.csv')\n",
    "#\n",
    "# aligned_df_exp_ligand_index = []\n",
    "# for i in range (39*2):\n",
    "#     aligned_df_exp_ligand_index = aligned_df_exp_ligand_index + list(np.array(ligand_in_expression_index)+45947*i)\n",
    "#\n",
    "# aligned_df_exp_ligand = df_expx.iloc[aligned_df_exp_ligand_index,]\n",
    "# aligned_df_exp_ligand.to_csv('E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\ligand_expression_aligned_yale.csv')\n",
    "\n",
    "##########################################\n",
    "aligned_df_exp_ligand= pd.read_csv('E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\ligand_expression_aligned_yale.csv',index_col =0 )\n",
    "aligned_df_exp_target= pd.read_csv('E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\target_expression_aligned_yale.csv',index_col =0 )\n",
    "aligned_ligand_target_matrix= pd.read_csv('E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\ligand_target_matrix_aligned_yale.csv',index_col =0 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "############################\n",
    "############################using cell_1 ligand expression fold change to predict cell_2 target gene expression fold change\n",
    "cell_type_pair_num = pd.read_csv('E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\IPF_control_cell_sampling_no_COPD_top_cell_types\\\\top_cell_types_37_IPF_vs_Control_cell_sampling_BDN_cell_type_pair_num.csv')\n",
    "prs_list = []\n",
    "for cell_pair_index in range (cell_type_pair_num.shape[0]): #### for each cell type pair\n",
    "    cell_type1,cell_type2 = cell_type_pair_num['cell_type1'][cell_pair_index],cell_type_pair_num['cell_type2'][cell_pair_index]\n",
    "    print (cell_type1,cell_type2)\n",
    "    target_expression_fold_control = np.array(aligned_df_exp_target[(aligned_df_exp_target['cell_type']==cell_type2)&(aligned_df_exp_target['disease']=='Control')]['average'])\n",
    "    target_expression_fold_disease = np.array(aligned_df_exp_target[(aligned_df_exp_target['cell_type']==cell_type2)&(aligned_df_exp_target['disease']=='IPF')]['average'])\n",
    "    target_name_list = np.array(aligned_df_exp_target[(aligned_df_exp_target['cell_type'] == cell_type2) & (aligned_df_exp_target['disease'] == 'IPF')]['gene'])\n",
    "    index_c = np.where(target_expression_fold_control!=0)[0]\n",
    "    index_d = np.where(target_expression_fold_disease!=0)[0]\n",
    "    index_cd = [i for i in index_c if i in index_d]\n",
    "    if len(index_cd)>0:\n",
    "        target_fold_change = target_expression_fold_disease[index_cd]/(target_expression_fold_control[index_cd])\n",
    "        target_name_filter_list = target_name_list[index_cd]\n",
    "        np.save('E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\IPF_control_cell_sampling_no_COPD_top_cell_types\\\\type_' + str(cell_pair_index) + 'target_filter.npy',target_name_filter_list)\n",
    "        # target_diff_change = target_expression_fold_disease - target_expression_fold_control\n",
    "        ligand_expression_fold_control = np.array(aligned_df_exp_ligand[(aligned_df_exp_ligand['cell_type']==cell_type1)&(aligned_df_exp_ligand['disease']=='Control')]['average'])\n",
    "        ligand_expression_fold_disease = np.array(aligned_df_exp_ligand[(aligned_df_exp_ligand['cell_type']==cell_type1)&(aligned_df_exp_ligand['disease']=='IPF')]['average'])\n",
    "        ligand_name_list = np.array(aligned_df_exp_ligand[(aligned_df_exp_ligand['cell_type'] == cell_type1) & (aligned_df_exp_ligand['disease'] == 'IPF')]['gene'])\n",
    "        index_c_ligand = np.where(ligand_expression_fold_control!=0)[0]\n",
    "        index_d_ligand = np.where(ligand_expression_fold_disease!=0)[0]\n",
    "        index_cd_ligand = [i for i in index_c_ligand if i in index_d_ligand]\n",
    "        if len(index_cd_ligand)>0:\n",
    "            ligand_fold_change = ligand_expression_fold_disease[index_cd_ligand]/ligand_expression_fold_control[index_cd_ligand]\n",
    "            ligand_name_filter_list = ligand_name_list[index_cd_ligand]\n",
    "            np.save('E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\IPF_control_cell_sampling_no_COPD_top_cell_types\\\\type_' + str(cell_pair_index) + 'ligand_filter.npy',ligand_name_filter_list)\n",
    "            # ligand_fold_change_filter = np.array([min(5,i) for i in ligand_fold_change])[:,np.newaxis]\n",
    "            ligand_fold_change =ligand_fold_change[:,np.newaxis]\n",
    "            # X_data = np.matmul(aligned_ligand_target_matrix,ligand_fold_change_filter)\n",
    "            X_data = np.array(aligned_ligand_target_matrix.iloc[index_cd,index_cd_ligand])*10**3\n",
    "            Y_data = np.log10(target_fold_change)\n",
    "            X_data = X_data*np.log10(ligand_fold_change[:,0])\n",
    "            X_data_temp = X_data+10**-10\n",
    "            ##########\n",
    "            ################################# regression model\n",
    "            import random\n",
    "            model_list = []\n",
    "            length_TF = Y_data.shape[0]\n",
    "            whole_data_TF = [i for i in range (length_TF)]\n",
    "            random.seed(1)\n",
    "            data_index = random.sample(whole_data_TF,length_TF)\n",
    "            X_datax = X_data_temp[data_index]\n",
    "            Y_datax = Y_data[data_index]\n",
    "            for test_indel in range (1,6):\n",
    "                print (test_indel)\n",
    "                test_TF = [i for i in range(int(np.ceil((test_indel - 1) * 0.2 * length_TF)), int(np.ceil(test_indel * 0.2 * length_TF)))]\n",
    "                train_TF = [i for i in whole_data_TF if i not in test_TF]\n",
    "                X_trainx = X_datax[train_TF]\n",
    "                Y_train = Y_datax[train_TF]\n",
    "                std_train = X_trainx.std(axis=0) ## training normalization using parameters from training dataset\n",
    "                mean_train = X_trainx.mean(axis=0)\n",
    "                X_train = (X_trainx - mean_train) / std_train  ###(X_IPF+10**-10)/(X_IPF.sum(axis=0)+10**-10)\n",
    "                # index3 = [i for i in index1 if i in index2]\n",
    "                X_testx = X_datax[test_TF]\n",
    "                X_test = (X_testx - mean_train) / std_train ## test normalization using parameters from training dataset\n",
    "                Y_test = Y_datax[test_TF]\n",
    "                t1 = time.time()\n",
    "                model = LassoCV(cv=3, alphas=2 ** np.arange(-10, 10, 0.5),max_iter=1000).fit(X_train, Y_train)\n",
    "                t_lasso_cv = time.time() - t1\n",
    "                print(t_lasso_cv)\n",
    "                # model.score(X_test,Y_test)\n",
    "                Y_predict = model.predict(X_test)\n",
    "                prsx = stats.pearsonr(Y_predict, Y_test)\n",
    "                prs_list.append(prsx[0])\n",
    "                model_list.append(model)\n",
    "                print (prs_list)\n",
    "                print (len(prs_list))\n",
    "                print (cell_type1, cell_type2)\n",
    "            for test_indel in range (1,6): ### five fold cross validation\n",
    "                model = model_list[test_indel-1]\n",
    "                m_log_alphas = -np.log2(model.alphas_)\n",
    "                fig = plt.figure()\n",
    "                # ymin, ymax = 2300, 3800\n",
    "                plt.plot(m_log_alphas, model.mse_path_, ':')\n",
    "                plt.plot(m_log_alphas, model.mse_path_.mean(axis=-1), 'k', label='Average across the folds', linewidth=2)\n",
    "                plt.axvline(-np.log2(model.alpha_), linestyle='--', color='k', label='alpha: CV estimate')\n",
    "                plt.legend()\n",
    "                plt.xlabel('-log2(alpha)')\n",
    "                plt.ylabel('Mean square error')\n",
    "                plt.title('Mean square error on each fold: coordinate descent ''(train time: %.2fs)' % t_lasso_cv)\n",
    "                plt.axis('tight')\n",
    "                plt.savefig('E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\IPF_control_cell_sampling_no_COPD_top_cell_types\\\\type_'+str(cell_pair_index)+'sy_IPF_CV_10_'+str(test_indel)+'.pdf')\n",
    "                del fig\n",
    "            with open('E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\IPF_control_cell_sampling_no_COPD_top_cell_types\\\\type_'+str(cell_pair_index)+'_sy_IPF_df3_model_list.pickle', 'wb') as f:\n",
    "                pickle.dump(model_list, f)\n",
    "        else:\n",
    "            prs_list = prs_list + [np.nan for kk in range(5)] ## failed\n",
    "    else:\n",
    "        prs_list = prs_list + [np.nan for kk in range (5)] ## failed\n",
    "\n",
    "np.save('E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\IPF_control_cell_sampling_no_COPD_top_cell_types\\\\type_prs_list.npy',np.array(prs_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "################## analysis between  the linear regression result AND Bayesian Network\n",
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "import matplotlib\n",
    "matplotlib.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "import random\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.linear_model import LassoCV, LassoLarsCV, LassoLarsIC\n",
    "from sklearn import datasets\n",
    "from scipy import stats\n",
    "import pickle\n",
    "\n",
    "######### linear regression result\n",
    "prs_list = np.load('E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\IPF_control_cell_sampling_no_COPD_top_cell_types\\\\type_prs_list.npy')\n",
    "######## Bayesian Network result\n",
    "cell_type_pair_num = pd.read_csv('E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\IPF_control_cell_sampling_no_COPD_top_cell_types\\\\top_cell_types_37_IPF_vs_Control_cell_sampling_BDN_cell_type_pair_num.csv')\n",
    "\n",
    "prs_reshape = prs_list.reshape(int(prs_list.shape[0]/5),5)\n",
    "prs_reshape_mean = prs_reshape.mean(axis=1)\n",
    "\n",
    "\n",
    "prs_reshape_mean_index = []\n",
    "df_prs = defaultdict(list)\n",
    "for threshold in range (21):\n",
    "    prs_reshape_mean_index = []\n",
    "    for i in range (prs_reshape_mean.shape[0]):\n",
    "        if not np.isnan(prs_reshape_mean[i]) and cell_type_pair_num['abs_diff'][i]>threshold: ## filter the result\n",
    "            prs_reshape_mean_index.append(i)\n",
    "    prs,pv = stats.pearsonr(prs_reshape_mean[prs_reshape_mean_index],np.log10(cell_type_pair_num['abs_diff'][prs_reshape_mean_index]))\n",
    "    print (prs)\n",
    "    df_prs['threshold'].append(threshold)\n",
    "    df_prs['pearson corr'].append(prs)\n",
    "\n",
    "df_prs_df = pd.DataFrame(df_prs)\n",
    "df_prs_df.to_csv('E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\IPF_control_cell_sampling_no_COPD_top_cell_types\\\\prs_threshold.csv')\n",
    "\n",
    "prs_reshape_mean_index = []\n",
    "for i in range (prs_reshape_mean.shape[0]):\n",
    "    if not np.isnan(prs_reshape_mean[i]) and cell_type_pair_num['abs_diff'][i]!=0:\n",
    "        prs_reshape_mean_index.append(i)\n",
    "\n",
    "stats.pearsonr(np.log10(prs_reshape_mean[prs_reshape_mean_index]), np.log10(cell_type_pair_num['abs_diff'][prs_reshape_mean_index]))\n",
    "\n",
    "\n",
    "df = defaultdict(list)\n",
    "df['log10(#edge)'] = list(np.log10(cell_type_pair_num['abs_diff'][prs_reshape_mean_index]))\n",
    "df['pearson corr for test'] = list(prs_reshape_mean[prs_reshape_mean_index])\n",
    "dfx = pd.DataFrame(df)\n",
    "fig = plt.figure()\n",
    "g=sns.jointplot('log10(#edge)','pearson corr for test',data = dfx,  kind=\"reg\")\n",
    "g.annotate(stats.spearmanr)\n",
    "plt.savefig('E:\\\\backup\\\\update_data_20201210\\\\update_data_jupyter\\\\IPF_control_cell_sampling_no_COPD_top_cell_types\\\\cell_type_LR_vs_BN.pdf')\n",
    "del fig"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
