# CINS
># 1, CINS
![](https://github.com/xiaoyeye/CINS/blob/main/CINS_overview.bmp)
Studies comparing single cell RNA-Seq (scRNA-Seq) data between conditions mainly focus on differences in the proportion of cell types or on differentially expressed genes. In many cases these differences are driven by changes in cell interactions which are challenging to infer without spatial information. To determine cell-cell interactions that differ between conditions we developed the Cell Interaction Network Inference (CINS) pipeline. CINS combines Bayesian network analysis with regression-based modeling to identify differential cell type interactions and the proteins that underlie them.  We tested CINS on a disease case control and on an aging human dataset. In both cases CINS correctly identifies cell type interactions and the ligands involved in these interactions. We performed additional mouse aging scRNA-Seq experiments which further support the interactions identified by CINS.
Overview of CINS. (a) Cell type information is discretized by learning Gaussian Mixture Model (GMM) for the fraction of cells of this type in each sample. (b) A Bayesian network (BN) is learned using the discretized cell abundance information. Bootstrapping is performed to identify significant interactions between cell types. (c) For pairs identified in the directed bootstrap BN analysis, a ligand-target regression (LTR) model is learned. In this model we use expression of ligands in the cell type with the outgoing edge to predict the expression of targets genes in the cell type with incoming edge. (d) Finally, LTR is used to select key ligands that underlie the cell-cell interactions identified in the BN. cell interaction.
